### PR TITLE
overlord/snapstate, snap: handle shared snap directories when installing/remove snaps with instance key

### PR DESCRIPTION
**Pull request by bboozzoo**
When installing snaps with instance key set, make sure to create common snap
directories that will be used as mountpoints when running the snap.
Specifically, when installing snap foo_bar, the following directories will be
created:
    
- /snap/foo  (bind mount of /snap/foo_bar)
- /var/snap/foo (bind mount of /var/snap/foo_bar)
    
Make sure that the directories are cleaned up when the snaps are removed and
there are no more snaps that may require them.
    
In order to keep the state lockign to minimum, the changeset extends the
snapstate backend operation with 2 additional methods for removing the base snap
directory (/snap/foo) and base snap data directory (/var/snap/foo).

### CODE REVIEW

File: overlord/snapstate/backend_test.go [hunk 1]
```diff
@@ -798,6 +800,26 @@ func (f *fakeSnappyBackend) RemoveSnapCommonData(info *snap.Info) error {
...
+		otherInstances: otherInstances,
+	})
+	return nil
+}
+
+func (f *fakeSnappyBackend) RemoveSnapDir(s snap.PlaceInfo, otherInstances bool) error {
+	f.ops = append(f.ops, fakeOp{
+		op:             "remove-snap-dir",
+		name:           s.InstanceName(),
+		path:           snap.BaseDir(snap.InstanceSnap(s.InstanceName())),
```
Comment (pedronis) [inline overlord/snapstate/backend_test.go]: I think it's ok to add SnapName to snap.PlaceInfo and simplify some code like this one

File: overlord/snapstate/handlers.go [hunk 2]
```diff
@@ -452,10 +452,29 @@ var (
 	mountPollInterval = 1 * time.Second
 )
 
+func otherSnapsLike(st *state.State, instanceName string) (bool, error) {
+	snapName, _ := snap.SplitInstanceName(instanceName)
+
+	all, err := All(st)
```
Comment (pedronis) [inline overlord/snapstate/handlers.go]: we don't need to unmarshal the SnapStates,  code like NumSnaps or SyncCookis would work,  maybe we should have  a rawSnaps internal helper and use it in those places, otoh the patter is just 4 lines

File: overlord/snapstate/handlers.go [hunk 3]
```diff
@@ -1204,6 +1261,11 @@ func (m *SnapManager) doDiscardSnap(t *state.Task, _ *tomb.Tomb) error {
 		}
 	}
 
+	otherInstances, err := otherSnapsLike(st, snapsup.InstanceName())
```
Comment (pedronis) [inline overlord/snapstate/handlers.go]: this needs to be computed only the if below case?

File: overlord/snapstate/handlers.go [hunk 4]
```diff
@@ -498,17 +517,29 @@ func (m *SnapManager) doMountSnap(t *state.Task, _ *tomb.Tomb) error {
 	}
 	if readInfoErr != nil {
 		if err := m.backend.UndoSetupSnap(snapsup.placeInfo(), snapType, pb); err != nil {
-			t.State().Lock()
```
Comment (pedronis) [inline overlord/snapstate/handlers.go]: this removal now looks wrong

File: overlord/snapstate/handlers.go [hunk 5]
```diff
@@ -498,17 +517,29 @@ func (m *SnapManager) doMountSnap(t *state.Task, _ *tomb.Tomb) error {
 	}
 	if readInfoErr != nil {
 		if err := m.backend.UndoSetupSnap(snapsup.placeInfo(), snapType, pb); err != nil {
```
Comment (pedronis) [inline overlord/snapstate/handlers.go]: wondering if it's worth having  a helper  

undoSetupSnap(st, placeInfo, snapType, pb)

for her are below or it's overkill

File: overlord/snapstate/handlers.go (hunk 4 above)
Comment (bboozzoo) [reply overlord/snapstate/handlers.go]: Fixed, added a test there too.

File: overlord/snapstate/backend/setup.go [hunk 6]
```diff
@@ -111,6 +114,23 @@ func (b Backend) RemoveSnapFiles(s snap.PlaceInfo, typ snap.Type, meter progress
...
 }
 
+func (b Backend) RemoveSnapDir(s snap.PlaceInfo, otherInstances bool) error {
+	mountDir := s.MountDir()
+
+	// try to remove the mount point base dir, failure is ok here
+	snapName, instanceKey := snap.SplitInstanceName(s.InstanceName())
+	if instanceKey != "" {
+		// always ok to remove instance specific one
+		os.Remove(filepath.Dir(mountDir))
```
Comment (mvo5) [inline overlord/snapstate/backend/setup.go]: We ignore failures here a comment about the rational for this would be great (same for the os.Remove a few lines below).

File: overlord/snapstate/backend/setup.go [hunk 7]
```diff
@@ -111,6 +114,23 @@ func (b Backend) RemoveSnapFiles(s snap.PlaceInfo, typ snap.Type, meter progress
 	return nil
 }
 
+func (b Backend) RemoveSnapDir(s snap.PlaceInfo, otherInstances bool) error {
```
Comment (mvo5) [inline overlord/snapstate/backend/setup.go]: I'm not a big fan of boolean arguments for functions. Let me show an example: `RemoveSnapDir(path, true)` is hard to parse unless the reader a) knows the function by heart (unlikely .) b) checks the function definition.  After reading the rest of the review I see this is never used "naked", always with a `otherInstances` var so its probably ok (even though I still don't like it that much :)

I also find the `otherInstances` not super clear. If true, does it mean "remove-snap-dir and all other instance dirs`? Or "remove-snap-dir but there are other instance dirs left". Would be great to find something that is more descriptive.

File: overlord/snapstate/backend/snapdata.go [hunk 8]
```diff
@@ -52,6 +52,22 @@ func (b Backend) RemoveSnapCommonData(snap *snap.Info) error {
 	return removeDirs(dirs)
 }
 
+// RemoveSnapDataDir removes base snap data directory
+func (b Backend) RemoveSnapDataDir(info *snap.Info, otherInstances bool) error {
```
Comment (mvo5) [inline overlord/snapstate/backend/snapdata.go]: Same comment as for the RemoveSnapDir.

File: overlord/snapstate/handlers.go [hunk 9]
```diff
@@ -452,10 +452,28 @@ var (
 	mountPollInterval = 1 * time.Second
 )
 
+func otherSnapsLike(st *state.State, instanceName string) (bool, error) {
```
Comment (mvo5) [inline overlord/snapstate/handlers.go]: I'm not super happy with the name but I can't think of anything that feels more catchy. I was wonderi if  `otherInstancesFor(st, instanceName)`, `hasOtherInstances(st, instanceName)` would be more descriptive. If not its probably fine but I would add a small docstring as the name does not give too much away.

File: snap/info.go [hunk 10]
```diff
@@ -38,9 +38,13 @@ import (
 
 // PlaceInfo offers all the information about where a snap and its data are located and exposed in the filesystem.
 type PlaceInfo interface {
-	// InstanceName returns the name of the snap.
+	// InstanceName returns the name of the snap decorated with instance
```
Comment (mvo5) [inline snap/info.go]: :heart: Thanks for improving this!

File: overlord/snapstate/backend/setup.go (hunk 7 above)
Comment (bboozzoo) [reply overlord/snapstate/backend/setup.go]: Tweaked the naming a bit

File: overlord/snapstate/handlers.go (hunk 9 above)
Comment (bboozzoo) [reply overlord/snapstate/handlers.go]: I actually like `hasOtherInstances()`.

File: overlord/snapstate/backend/snapdata.go [hunk 11]
```diff
@@ -72,9 +92,6 @@ func removeDirs(dirs []string) error {
 		if err := os.RemoveAll(dir); err != nil {
 			return err
 		}
-
-		// Attempt to remove the parent directory as well (ignore any failure)
-		os.Remove(filepath.Dir(dir))
```
Comment (mvo5) [inline overlord/snapstate/backend/snapdata.go]: Iirc this was used so that removing the last /snap/foo/123 revision would result in /snap/foo going away. Is this still accounted for somewhere? It sounds like we should have a spread test for this (to ensure that dirs are cleaned up if no snaps are installed anymore). Not sure if we have one right now.

Comment (bboozzoo) [reply overlord/snapstate/backend/snapdata.go]: There's a spread test included in the PR. It does installation, removal and checks whether relevant directories exist. Although it already uses snaps with and without instance key, there's no case for instance-key-less snap being the sole one installed. I can extend it as needed.

Comment (bboozzoo) [reply overlord/snapstate/backend/snapdata.go]: Extended the spread test to cover this too.

File: overlord/snapstate/backend/setup.go [hunk 12]
```diff
@@ -57,6 +57,13 @@ func (b Backend) SetupSnap(snapFilePath, instanceName string, sideInfo *snap.Sid
 		return snapType, err
 	}
 
+	if s.InstanceKey != "" {
+		err := os.MkdirAll(snap.BaseDir(s.SnapName()), 0755)
```
Comment (pedronis) [inline overlord/snapstate/backend/setup.go]: do we need to remove this in the defer above in some case?

File: overlord/snapstate/backend/snapdata.go (hunk 11 above)
Comment (pedronis) [reply overlord/snapstate/backend/snapdata.go]: are we sure this accounted for the undo cases,  we added explicit new Remove* calls for UndoSetupSnap but I don't see the equivalent for UndoCopySnapData (which would use removeDirs doing this remove before)

Comment (pedronis) [reply overlord/snapstate/backend/snapdata.go]: also this bit was used by RemoveSnapCommonData  that didn't go a new remove

File: overlord/snapstate/backend/setup.go (hunk 12 above)
Comment (bboozzoo) [reply overlord/snapstate/backend/setup.go]: This has to be cleaned up by the caller in snapstate. Here we do not know if the directory is perhaps used by another snap once we get to the cleanup part.

File: overlord/snapstate/backend/snapdata.go (hunk 11 above)
Comment (bboozzoo) [reply overlord/snapstate/backend/snapdata.go]: As you noted the cleanup was missing from undoCopySnapData and doCopySnapData error path. Fixed that and added a spread test to cover this too.

File: overlord/snapstate/backend/setup.go (hunk 12 above)
Comment (pedronis) [reply overlord/snapstate/backend/setup.go]: you mean is covered by the new code in doMountSnap error handling?
